### PR TITLE
Enable CosmosDB OAuth support

### PR DESCRIPTION
**Pull request by JasonYeMSFT**
resolves: https://github.com/microsoft/vscode-cosmosdb/issues/2205

This PR duplicates some internal implementation of the auth package as the workaround to its limited export.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1069,6 +1069,11 @@
                     "type": "integer",
                     "default": 8081,
                     "description": "Port to use when connecting to a CosmosDB Emulator instance"
+                },
+                "azureDatabases.testCosmosAuth": {
```
Comment (JasonYeMSFT) [inline package.json:1073]: I intend to keep this setting until the testing is done. This allows us to force testing the auth credential and compare with key credentials without going back and forth toggling the disableLocalAuth on the resource.

File: src/commands/api/DatabaseAccountTreeItemInternal.ts [hunk 2]
```diff
@@ -57,10 +58,15 @@ export class DatabaseAccountTreeItemInternal implements DatabaseAccountTreeItem
 
     public get docDBData(): { masterKey: string; documentEndpoint: string; } | undefined {
         if (this._accountNode instanceof DocDBAccountTreeItemBase) {
-            return {
-                documentEndpoint: this._accountNode.root.endpoint,
-                masterKey: this._accountNode.root.masterKey
-            };
+            const keyCred = this._accountNode.root.credentials.filter((cred): cred is CosmosDBKeyCredential => cred.type === 'key')[0];
```
Comment (MRayermannMSFT) [inline src/commands/api/DatabaseAccountTreeItemInternal.ts:61]: `.at(0)`

File: src/docdb/commands/connectNoSqlContainer.ts [hunk 3]
```diff
@@ -7,16 +7,19 @@ import { IActionContext } from "@microsoft/vscode-azext-utils";
 import { KeyValueStore } from "../../KeyValueStore";
 import { ext } from "../../extensionVariables";
 import { NoSqlQueryConnection, noSqlQueryConnectionKey } from "../NoSqlCodeLensProvider";
+import { CosmosDBKeyCredential } from "../getCosmosClient";
 import { DocDBCollectionTreeItem } from "../tree/DocDBCollectionTreeItem";
 import { pickDocDBAccount } from "./pickDocDBAccount";
 
 export function setConnectedNoSqlContainer(node: DocDBCollectionTreeItem): void {
+    const root = node.root;
+    const keyCred: CosmosDBKeyCredential | undefined = root.credentials.filter((cred): cred is CosmosDBKeyCredential => cred.type === "key")[0];
```
Comment (MRayermannMSFT) [inline src/docdb/commands/connectNoSqlContainer.ts:16]: .at(0)

File: src/docdb/getCosmosClient.ts [hunk 4]
```diff
@@ -4,15 +4,70 @@
...
+
+export type CosmosDBCredential = CosmosDBKeyCredential | CosmosDBAuthCredential;
+
+/**
+ * Duplicated from @microsoft/vscode-azext-azureauth
+ * @todo: Use the subscription client once it supports customizing scopes.
+ */
+export async function getSessionFromVSCode(scopes: string | string[], options?: vscode.AuthenticationGetSessionOptions): Promise<vscode.AuthenticationSession | undefined> {
+    const scopesArray = typeof scopes === "string" ? [scopes] : scopes;
+    return await vscode.authentication.getSession(getConfiguredAuthProviderId(), scopesArray, options);
```
Comment (MRayermannMSFT) [inline src/docdb/getCosmosClient.ts:30]: Nit: use the internal auth module:
image

File: src/docdb/getCosmosClient.ts [hunk 5]
```diff
@@ -4,15 +4,70 @@
...
+    endpoint: string,
+    credentials: CosmosDBCredential[],
+    isEmulator: boolean | undefined
+): CosmosClient {
     const vscodeStrictSSL: boolean | undefined = vscode.workspace.getConfiguration().get<boolean>(ext.settingsKeys.vsCode.proxyStrictSSL);
     const enableEndpointDiscovery: boolean | undefined = vscode.workspace.getConfiguration().get<boolean>(ext.settingsKeys.enableEndpointDiscovery);
     const connectionPolicy = { enableEndpointDiscovery: (enableEndpointDiscovery === undefined) ? true : enableEndpointDiscovery };
-    return new CosmosClient({ endpoint, key, userAgentSuffix: appendExtensionUserAgent(), agent: new https.Agent({ rejectUnauthorized: isEmulator ? !isEmulator : vscodeStrictSSL }), connectionPolicy: connectionPolicy });
 
+    const keyCred = credentials.filter((cred): cred is CosmosDBKeyCredential => cred.type === "key")[0];
```
Comment (MRayermannMSFT) [inline src/docdb/getCosmosClient.ts:42]: 🙃

File: src/docdb/tree/DocDBAccountTreeItemBase.ts [hunk 6]
```diff
@@ -24,22 +24,38 @@ export abstract class DocDBAccountTreeItemBase extends DocDBTreeItemBase<Databas
...
-        this.valuesToMask.push(id, endpoint, masterKey);
+        const keys = credentials
+            .map((cred) => cred.type === "key" ? cred.key : undefined)
+            .filter((value): value is string => value !== undefined);
+        this.valuesToMask.push(id, endpoint, ...keys);
     }
 
     public get connectionString(): string {
-        return `AccountEndpoint=${this.root.endpoint};AccountKey=${this.root.masterKey}`;
+        const firstKey = this.root.credentials.filter((cred): cred is CosmosDBKeyCredential => cred.type === "key")[0];
```
Comment (MRayermannMSFT) [inline src/docdb/tree/DocDBAccountTreeItemBase.ts:53]: 🧐

**Comment by MRayermannMSFT** [comment]
Mostly lgtm.

File: src/commands/api/DatabaseAccountTreeItemInternal.ts (hunk 2 above)
Comment (JasonYeMSFT) [reply src/commands/api/DatabaseAccountTreeItemInternal.ts:61]: The 'lib' in tsconfig is too old to support .at. Changing it results in multiple build error to surface. I'll skip using .at in this PR.

File: src/docdb/commands/connectNoSqlContainer.ts (hunk 3 above)
Comment (JasonYeMSFT) [reply src/docdb/commands/connectNoSqlContainer.ts:16]: The 'lib' in tsconfig is too old to support .at. Changing it results in multiple build error to surface. I'll skip using .at in this PR.

File: src/docdb/getCosmosClient.ts (hunk 5 above)
Comment (JasonYeMSFT) [reply src/docdb/getCosmosClient.ts:42]: The 'lib' in tsconfig is too old to support .at. Changing it results in multiple build error to surface. I'll skip using .at in this PR.

File: src/docdb/tree/DocDBAccountTreeItemBase.ts (hunk 6 above)
Comment (JasonYeMSFT) [reply src/docdb/tree/DocDBAccountTreeItemBase.ts:53]: The 'lib' in tsconfig is too old to support .at. Changing it results in multiple build error to surface. I'll skip using .at in this PR.

Comment (JasonYeMSFT) [reply src/docdb/tree/DocDBAccountTreeItemBase.ts]: Good catch btw.

**Review by alexweininger — DISMISSED** [review]

**Review by alexweininger — APPROVED** [review]

**Pull request merged**